### DISCUSSION TITLE
Fetch a release note from a source repository for Docker dependencies

### DISCUSSION THREAD

**Discussion by grezar** [renovatebot/renovate#15300] [Request Help]
### What would you like Renovate to be able to do?

Renovate can refer to the image registry of the container and update the image tag in the docker.
However, since it only refers to the image registry, it cannot fetch release notes from the VCS repo, and it is not possible to include such information in the PR.
I think this is because the container registry usually does not have mapping information to the VCS repository that the image is based on.

As discussed here https://github.com/renovatebot/renovate/issues/2395, it is currently possible to refer to the VCS repository by including a label in the image manifest, but it is inconvenient to not be able to set this freely for other people's projects. 
Therefore, I think it would be useful if there was a feature to set up the VCS repository in the config of Renovate itself, but what do you think about such a manual mapping feature?

### If you have any ideas on how this should be implemented, please tell us here.

Nothing yet

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
For now we support this in code: https://github.com/renovatebot/renovate/blob/29dde2f6941c7f102bfc06cadb05e0e35e9204e9/lib/datasource/metadata.ts#L69

Allowing users to override values in config would mess with our caching and potentially cause one user's mistake or malicious config to affect others using the same package

**Comment by viceice** [comment]
We also support docker label `org.opencontainers.image.source` to point to source

https://github.com/renovatebot/renovate/blob/29dde2f6941c7f102bfc06cadb05e0e35e9204e9/lib/datasource/docker/index.ts#L267-L269

**Comment by grezar** [comment]
@rarkins 
I see.
I didn't take into account the cache-related issues 😮 
So, for now, addding a new mapping into that variable is the way to do this, right?
If so, I'm willing to create a PR for it 👍🏼

**Comment by viceice** [comment]
but first step would be to get the image author to add the docker label instead

**Comment by grezar** [comment]
@viceice 
Hmm. Okay.
Since I have several docker dependencies, it's a bit tedious to ask authors to add the label, but yeah, I will.

**Comment by scrocquesel** [comment]
> We also support docker label `org.opencontainers.image.source` to point to source
> 
> https://github.com/renovatebot/renovate/blob/29dde2f6941c7f102bfc06cadb05e0e35e9204e9/lib/datasource/docker/index.ts#L267-L269

Shouldn't this be documented ? I guess adding a readme.md in /lib/datasource/docker saying it support OCI Image annotation specification. And BTW, also support the not so deprecated label schema (`org.label-schema.vcs-url`) ?

**Comment by rarkins** [comment]
We don't document how we use every field of every datasource we support, but if you think it would help others then you can create a readme.md in https://github.com/renovatebot/renovate/tree/main/lib/datasource/docker. Here is an example of the readme file from another datasource: https://github.com/renovatebot/renovate/blob/main/lib/datasource/gitlab-packages/readme.md which then gets imported into this doc: https://docs.renovatebot.com/modules/datasource/#gitlab-packages-datasource

**Comment by viceice** [comment]
We've also an undocumented /unsupported feature to override `sourceUrl` via package rule 🙃 

```json
    {
      "matchPackageNames": ["jenkins"],
      "matchDatasources": ["docker"],
      "sourceUrl": "https://jenkins.io/changelog-stable"
    },
```

https://github.com/visualon/renovate-config/blob/main/meta.json

**Comment by OuesFa** [comment]
> For now we support this in code:
> 
> https://github.com/renovatebot/renovate/blob/29dde2f6941c7f102bfc06cadb05e0e35e9204e9/lib/datasource/metadata.ts#L69
> 
> Allowing users to override values in config would mess with our caching and potentially cause one user's mistake or malicious config to affect others using the same package

@rarkins Is it possible to freely update this configuration if we self-host renovate?

**Comment by rarkins** [comment]
If you self-host, you can use packageRules with `sourceUrl` strings to add them.

**Comment by OuesFa** [comment]
Thanks @rarkins 
I can't really figure out how to indicate the mapping properly using the `packageRules` config.

Here is the config I'm trying
```json
  "packageRules": [
    {
      "matchSourceUrls": [
        "https://github.com/elastic/cloud-on-k8s"
      ],
      "groupName": "ES packages",
      "matchPaths": [
        "dbre/elastic/**"
      ]
    }
  ]
```
Here is my basic Dockerfile located within the path indicated above
```
FROM docker.elastic.co/eck/eck-operator:1.7.0
```

In the PR opened by renovate to update the package, there is no release notes. 

I'm wondering if it's a format issue, in the `manualSourceUrls` field I can see that the mapping should look like this 
```
docker: {
    '$packageName': '$githubRepoUrl',
}
```
WDYT? 

FYI I also opened an issue to add the OCI label in the upstream docker images https://github.com/elastic/cloud-on-k8s/issues/5611

**Comment by rarkins** [comment]
`matchPackageNames=["docker.elastic.co/eck/eck-operator"]`,
`sourceUrl=https://github.com/elastic/cloud-on-k8s`

**Reply by OuesFa** [reply]
Thanks for your quick answer @rarkins!
There is no sourceUrl field in the `packageRules` conf.
I tried with `matchUrlSources` config.

I got this error `Repository has invalid config` `Invalid configuration option: packageRules[0].matchSourceUrls`
Here are the details (I'm omitting some unnecessary fields)
```json
{"name":"renovate","hostname": [],"error":{"validationSource":"renovate.json5","validationError":"The renovate configuration file contains some invalid settings","validationMessage":"Invalid configuration option: packageRules[0].matchSourceUrls","message":"config-validation","stack":"Error: config-validation\n    at mergeRenovateConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/merge.ts:183:19)\n    at getRepoConfig (/usr/src/app/node_modules/renovate/lib/workers/repository/init/config.ts:16:12)\n    at initRepo (/usr/src/app/node_modules/renovate/lib/workers/repository/init/index.ts:31:12)\n    at Object.renovateRepository (/usr/src/app/node_modules/renovate/lib/workers/repository/index.ts:40:14)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:150:7)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"},"msg":"Repository has invalid config","time":"2022-04-25T13:38:04.806Z","v":0} 
```

**Reply by rarkins** [reply]
Use `sourceUrl` like above

**Reply by OuesFa** [reply]
I seem to get the same error when using `sourceUrl`
"error":{"validationSource":"renovate.json5","validationError":"The renovate configuration file contains some invalid settings","validationMessage":"Invalid configuration option: packageRules[0].sourceUrl","message":"config-validation","stack":"Error: config-validation

Here is my conf 
```jsoon
    "packageRules": [
      {
        "allowedPostUpgradeCommands": [
          ".*"
        ],
        "allowPostUpgradeCommandTemplating": true,
        "postUpgradeTasks": {
          "commands": [
            "some command"
          ],
          "fileFilters": [
            "**/Dockerfile"
          ],
          "executionMode": "update"
        },
        "groupName": "DBRE packages",
        "matchPackageNames": [
          "docker.elastic.co/eck/eck-operator"
        ],
        "sourceUrl": "https://github.com/elastic/cloud-on-k8s",
        "matchPaths": [
          "dbre/**",
          "charts/cp-schema-registry/**"
        ]
      },
```

**Reply by rarkins** [reply]
It can only be added to your config.js file on the server side

**Reply by OuesFa** [reply]
When put on server side config it works like a charm thanks! 
How can I make it work for multiple packages please? 
As of now, from what I understand it iterates on the `matchPackageNames` array and matches every package with the url in the `sourceUrl`. However, in my case I will need to match different packages to different urls.

**Reply by viceice** [reply]
yes, I'm using a shared preset. 🙃

https://github.com/visualon/renovate-config/blob/main/meta.json

**Reply by OuesFa** [reply]
@viceice thanks! 
Presets can be used as server side configs? Could you point me where you're using it?

**Reply by viceice** [reply]
it's nested to my default shared preset, all is in that repo.
